pkg/core/chain/chain.go

```diff
@@ -143,111 +211,129 @@ func New(ctx context.Context, db database.DB, eventBus *eventbus.EventBus, rpcBu
+			c.lock.Unlock()
 			return err
 		}
 
-		// Otherwise, accept the block directly.
-		// NOTE: The error is not checked for relevance here, since this is
-		// already done in the Agreement. The Agreement has direct access
-		// to `CatchBlockChan` during consensus, and will be the first one
-		// to receive such messages. There, the error is checked for relevance
-		// by comparing the block height to the current round. This makes sure
-		// that an irrelevant error is never bubbled up to this function, and
-		// thus, we don't need to check it here.
-		if !block.IsEmpty() && block.Header.Height == c.CurrentHeight()+1 {
-			if err = c.AcceptSuccessiveBlock(block); err != nil {
+		c.lock.Unlock()
+
+		select {
+		case candidate := <-winnerChan:
+			c.lock.Lock()
+
+			block, err := candidate.Blk, candidate.Err
+			if err != nil {
+				// Most likely a context cancellation, but could also be a reaching
+				// of maximum steps.
+				c.lock.Unlock()
 				return err
 			}
+
+			if !block.IsEmpty() && block.Header.Height == c.CurrentHeight()+1 {
+				if err = c.AcceptSuccessiveBlock(block); err != nil {
+					c.lock.Unlock()
+					return err
+				}
+			}
+
+			c.lock.Unlock()
+		case <-c.stopConsensusChan:
+			return nil
 		}
 	}
 }
 
-func (c *Chain) produceBlock(ctx context.Context) (winner consensus.Results) {
+func (c *Chain) produceBlock(ctx context.Context, winnerChan chan consensus.Results) error {
 	ru := c.GetRoundUpdate()
 
 	if c.loop != nil {
 		scr, agr, err := c.loop.CreateStateMachine(c.db, config.ConsensusTimeOut, c.VerifyCandidateBlock, c.CatchBlockChan)
 		if err != nil {
-			// TODO: errors should be handled by the caller
 			log.WithError(err).Error("could not create consensus state machine")
-
-			winner.Err = err
-			return winner
+			return err
 		}
 
-		return c.loop.Spin(ctx, scr, agr, ru)
+		go func() {
+			winnerChan <- c.loop.Spin(ctx, scr, agr, ru)
+		}()
 	}
 
-	for {
-		select {
-		case r := <-c.CatchBlockChan:
-			if r.Blk.Header != nil && r.Blk.Header.Height != ru.Round {
-				continue
-			}
-
-			winner = r
-			return
-		case <-ctx.Done():
-			return
-		}
-	}
+	return errors.New("no consensus loop present")
 }
 
-// ProcessSucceedingBlock will handle blocks incoming from the network,
-// which directly succeed the known chain tip.
-func (c *Chain) ProcessSucceedingBlock(blk block.Block) error {
-	log.WithField("height", blk.Header.Height).Trace("received succeeding block")
-
-	select {
-	case c.CatchBlockChan <- consensus.Results{Blk: blk, Err: nil}:
-	default:
-		return errors.New("block discarded")
+func (c *Chain) ProcessSuccessiveBlock(blk block.Block) error {
+	// If we can not stop the consensus, it means that the ProduceBlock goroutine
+	// is waiting to take ownership of the mutex. In this scenario, it means
+	// that consensus has converged on a block. We will give precendence to
+	// self-finalized blocks over network-received ones, so we relinquish control
+	// over the mutex in case of an error.
```

Comment:
The description does not really reflect what is in the code. I cannot easily see where the control over the mutex gets relinquished as a consequence of errors returned by `StopBlockProduction`. If a refactoring leading to more explicit code is not easy to obtain, at the very least the comment should be more specific and detailed about side effects in the mutex handling (which I suppose should be in the synchronization state-machine, but I am not sure)